examples/api/lib/material/text_form_field/text_form_field.2.dart

```diff
@@ -0,0 +1,143 @@
+
+void main() => runApp(const TextFormFieldExampleApp());
+
+class TextFormFieldExampleApp extends StatelessWidget {
+  const TextFormFieldExampleApp({super.key});
+
+  @override
+  Widget build(BuildContext context) {
+    return const MaterialApp(
+      home: TextFormFieldExample(),
+    );
+  }
+}
+
+class TextFormFieldExample extends StatefulWidget {
+  const TextFormFieldExample({super.key});
+
+  @override
+  State<TextFormFieldExample> createState() => _TextFormFieldExampleState();
+}
+
+class _TextFormFieldExampleState extends State<TextFormFieldExample> {
+  late final TextEditingController controller;
+  final GlobalKey<FormState> formKey = GlobalKey<FormState>();
+  String? forcedErrorText;
+  bool isLoading = false;
+  // username as a key, error text as value.
+  final Map<String, String> cachedServerErrors = <String, String>{};
+
+  @override
+  void initState() {
+    super.initState();
+    controller = TextEditingController();
+  }
+
+  @override
+  void dispose() {
+    controller.dispose();
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Material(
+      child: Padding(
+        padding: const EdgeInsets.symmetric(horizontal: 24),
+        child: Center(
+          child: Form(
+            key: formKey,
+            child: Column(
+              mainAxisAlignment: MainAxisAlignment.center,
+              children: <Widget>[
+                TextFormField(
+                  forceErrorText: forcedErrorText,
+                  controller: controller,
+                  decoration: const InputDecoration(
+                    hintText: 'Please write a username',
+                  ),
+                  validator: (String? value) {
+                    if (value == null || value.isEmpty) {
+                      return 'This field is required';
+                    }
+                    if (value.length != value.replaceAll(' ', '').length) {
+                      return 'Username must not contains any spaces';
+                    }
+                    if (int.tryParse(value[0]) != null) {
+                      return 'username must not start with a number';
+                    }
+                    if (value.length <= 2) {
+                      return 'username should be at least 3 characters long';
+                    }
+                    return null;
+                  },
+                  onChanged: (String value) {
+                    if (forcedErrorText != null && !cachedServerErrors.containsKey(value)) {
+                      setState(() {
+                        forcedErrorText = null;
+                      });
+                    }
+                  },
+                ),
+                const SizedBox(height: 40),
+                if (isLoading)
+                  const CircularProgressIndicator()
+                else
+                  TextButton(
+                    child: const Text('Save'),
+                    onPressed: () async {
```

Comment:
Another good candidate to move into a method above in order to make this more readable, if you think it would help.